Introduce support for CMSIS RTOS v1 APIs

rgundi opened the pull request:
This PR mainly introduces support for thread and timer management APIs for CMSIS RTOS V1.

rgundi committed:
lib/cmsis_rtos_v1: Implement support for thread APIs

These APIs allow defining, creating and controlling of thread
related functionalities.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
CMSIS RTOS V1: Introduce CMSIS RTOS V1 API header file

CMSIS RTOS API provides a generic RTOS interface for embedded
processors (actually for Cortex-M processors but are generic
enough to be used elsewhere). This header file is for V1 version.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of Thread APIs

Some test programs to illustrate the usage of thread APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of Timer APIs

Some test programs to illustrate the usage of timer APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v1: Implement support for osDelay

This API is used to specify delay in ms. This is analogous
to k_sleep in the kernel.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v1: Implement support for osKernelSysTick

This API is used to fetch the kernel system timer as 32-bit value.
This is analogous to k_cycle_get_32 in the kernel.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v1: Implement support for timer APIs

These APIs provide the support of virtual timers. All timers
can be started, restarted, or stopped. Timers can be configured
as one-shot or periodic.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@nashif cmsis_os.h needs to be excluded from checkpatch. Let me know how to do it.

andrewboie commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,242 @@
+}
+
+/* Memory pool for cmsis thread space */
+K_MEM_POOL_DEFINE(cmsis_thread_pool, sizeof(struct cmsis_thread),
```
consider using k_mem_slab APIs instead, they are specifically designed for pools of fixed-size objects

rgundi replied:
sure. will check.

andrewboie commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,242 @@
+		return NULL;
+	}
+
+	(*(thread_def->used_inst_ptr))++;
```
don't see any synchronization, could we use an atomic operation for the test-and-increment

rgundi replied:
good point. will check.

andrewboie commented on include/cmsis_rtos_v1/cmsis_os.h:
```diff
@@ -0,0 +1,703 @@
+  uint32_t               instances;    ///< maximum number of instances of that thread function
+  uint32_t               stacksize;    ///< stack size requirements in bytes; 0 is default stack size
+  uint32_t          *used_inst_ptr;    ///< indicates the number of instances used up
+  k_thread_stack_t   (*stack_mem)[CONFIG_MAX_STACK_SIZE_THREAD];    ///< pointer to array of stack memory
```
don't do this.
Use K_THREAD_STACK_MEMBER

rgundi replied:
Well.. From the documentation, i see that K_THREAD_STACK_MEMBER should be used for stacks embedded within other data structures for memory protection reasons. Here, "stack_mem" is pointing to an "array of stacks" i.e. of the type stack_mem[i][j]. An array of stacks is needed because CMSIS RTOS V1 has this somewhat weird requirement of "instances" while creating a thread. Check the below link to get an idea of what i am trying to say. http://www.keil.com/pack/doc/CMSIS/RTOS/html/group__CMSIS__RTOS__ThreadMgmt.html#gaf0c7c6b5e09f8be198312144b5c9e453

andrewboie commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,242 @@
+	cm_thread->block.id = block.id;
+
+	tid = k_thread_create(&cm_thread->thread,
+			thread_def->stack_mem[*(thread_def->used_inst_ptr) - 1],
```
should just be thread_def->stack_mem if you defined it with K_THREAD_STACK_MEMBER

rgundi commented:
Please ignore the sanitycheck failures for the time being. I am fixing them but the reviewers should be able to review the code in the meantime.

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to verify Thread instances

Some tests to verify the implementation of thread instances.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

andyross commented on lib/cmsis_rtos_v1/cmsis_kernel.c:
```diff
@@ -0,0 +1,19 @@
+uint32_t osKernelSysTick(void)
+{
+	return k_cycle_get_32();
+}
```
A whole source file with copyright header and two (!) include files just for a single one-line function wrapper.  Seriously? :)

Can't you fold this together with the other files?

rgundi replied:
This will have further additions (at least 3 more functions).

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+	cm_thread = CONTAINER_OF(thread_id, struct cmsis_thread, thread);
+
+	worker.thrd = cm_thread;
+	k_work_submit(&worker.work.work);
```
I see what you're doing here, but this trick isn't as safe as it looks.  You want to guarantee that the thread is truly finished executing before you do the free of its thread struct.  But a work queue doesn't guarantee that.  If someone has played with priorities (or if you're on an SMP system, though AFAIK there are no plans for CMSIS on Cortex A) you could theoretically find the work queue executing before the thread is finished.

Getting this right is going to be tricky.  Some combination of locking interrupts, giving a semaphore and then calling k_thread_abort(k_thread_current()) is probably going to be needed.

rgundi replied:
Yes. We too have noticed this. In fact, I have sent a mail on this (you are included). IMO, this can be addressed in 2 ways.
a.	Decrease the priority of the work queue thread (default is -1) to that of the Idle thread. Then, when a thread is aborted, it’s state is set accordingly. The work queue thread can look at this state and then free the corresponding memory occupied by the thread.
b.	Have a new thread of Idle priority which will do the above job. Something like a clean-up thread.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+		 "invalid stack size\n");
+
+	if (_is_in_isr()) {
+		return NULL;
```
I don't get it.  Why can't you create a thread in an ISR?  Pretty sure it should work, no?  Do we document it otherwise?

rgundi replied:
This is what the spec says. Not allowed as per spec.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+
+	if (*(thread_def->used_inst_ptr) >= thread_def->instances) {
+		return NULL;
+	}
```
Are those known runtime failure cases that need to be handled? This looks like you're just trying to validate the user-supplied parameters.  Zephyr style is to do that with an ASSERT and leave production runtime builds free of cruft.

rgundi replied:
thanks for pointing out. Will use ASSERT.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+
+	if (k_mem_pool_alloc(&cmsis_thread_pool, &block,
+			     sizeof(struct cmsis_thread), 100) != 0) {
+		/* Insufficient resource to create thread */
```
/* Yeah.  That's what a failed call to alloc means */

(obvious comment is obvious, remove)

rgundi replied:
will remove.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+		return NULL;
+	}
+
+	atomic_inc((atomic_t *)(thread_def->used_inst_ptr));
```
What other threads is this atomic call protecting against?  The thread_def struct is specific to the thread being created, and this is the create call, right?

rgundi replied:
Well.. this is mostly a paranoid protection.  It can so happen that the thread calling this create call might get swapped out due to time slice expiration and the new thread may do the same create call. We will run into issues in such a case.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+{
+	if (_is_in_isr()) {
+		return NULL;
+	}
```
That ISR check seems senseless.  Does CMSIS require that this return NULL in an ISR?  If not, remove.

andyross replied:
FWIW: when they say "cannot be called" they mean "don't do that" not "must signal an error".  It's not uncompliant for a CMSIS implementation, as e.g. an extension, to return (like Zephyr does!) the thread ID of the interrupted thread here.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+
+	if (_is_in_isr()) {
+		return osPriorityError;
+	}
```
Again, seems like error checking isn't really appropriate.  Remove unless required by the spec.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
+
```
More error checking to replace with ASSERTs

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+	}
+
+	z_prio = cmsis_to_zephyr_priority(priority);
+	k_thread_priority_set(thread, z_prio);
```
Would save two lines and read more cleanly to write

`k_thread_priority_set(thread, cmsis_to_zephyr_priority(priority));`

rgundi replied:
will do.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
```
ASSERT please

rgundi replied:
Required by spec.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+	 */
+	cm_thread = CONTAINER_OF(thread_id, struct cmsis_thread, thread);
+	worker.thrd = cm_thread;
+	k_work_submit(&worker.work.work);
```
Same comment.  This isn't really a robust way of doing this.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,246 @@
+{
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
```
ASSERT

rgundi replied:
Required by spec.

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.h:
```diff
@@ -0,0 +1,19 @@
+	struct k_thread thread;
+};
+
+#endif
```
A whole header with a copyright declaration, license ID, include guards and two include files (one of which is completely superfluous, btw: there are no symbols from cmsis_os referenced here) just to declare a two-field struct.

Is this header an official CMSIS thing that external apps will expect to find?  If so, then I guess that's unavoidable.  But would be nice to fold this together with other files.

rgundi replied:
Thanks for pointing this out. It was originally created to hold a little more but then got reduced to only 1 struct. Will move this to the corresponding .c file.

andyross commented on lib/cmsis_rtos_v1/cmsis_wait.c:
```diff
@@ -0,0 +1,27 @@
+	}
+
+	k_sleep(delay_ms);
+	return osEventTimeout;
```
Another whole file just for a tiny wrapper.  Fold with something else.

rgundi replied:
This may have at least one more function coming up. Will take a decision later depending upon how big that is going to be.

andyross commented on lib/cmsis_rtos_v1/cmsis_timer.c:
```diff
@@ -0,0 +1,126 @@
+
+	timer = (struct timer_obj *)ztimer;
+
+	(timer->callback_function)(timer->arg);
```
Use CONTAINER_OF() here.  Yeah, I see you put the k_timer at the beginning of the struct.  That's a clever footgun.  Do it right :)

rgundi replied:
Will do.

andyross commented on lib/cmsis_rtos_v1/cmsis_timer.c:
```diff
@@ -0,0 +1,126 @@
+
+	if (timer->status == ACTIVE) {
+		timer->status = NOT_ACTIVE;
+		k_timer_stop(&timer->ztimer);
```
Do you really need that status field?  I'm all but certain that the k_timer API already has the behavior you're reimplementing here.  Stopping an inactive timer is a silent noop already.

rgundi replied:
Thanks for pointing out that the k_timer API already has this behavior. However, calling _abort_timeout needs me to include quite a few other header files which are otherwise unnecessary. Also, per CMSIS documentation, stopping an inactive timer is an error. Hence, i prefer keeping the status field.

andyross commented on lib/cmsis_rtos_v1/cmsis_timer.c:
```diff
@@ -0,0 +1,126 @@
+	}
+
+	if (timer->status == NOT_ACTIVE) {
+		return osErrorResource;
```
Is this a documented error?  That sounds wrong to me.  Normal race conditions vs. timer interrupts mean that it's basically always possible to "stop" a timer after it's fired.  It should be a noop, not an error, I suspect.  Can you check the CMSIS documentation?

rgundi replied:
Yes. This is a documented error in CMSIS. They basically want to catch a situation where a timer is being stopped when it is not even started.

andyross commented on lib/cmsis_rtos_v1/cmsis_timer.c:
```diff
@@ -0,0 +1,126 @@
+
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
```
More error checking that should be ASSERTs.

rgundi replied:
Required by spec.

andyross commented on lib/cmsis_rtos_v1/cmsis_timer.c:
```diff
@@ -0,0 +1,126 @@
+	if (_is_in_isr()) {
+		return osErrorISR;
+	}
+
```
ASSERTs

rgundi replied:
Required by spec.

andyross reviewed (changes requested):
Mostly style whining, but the racy thread cleanup needs a better story than it gets here.   Either we have to do this right, prove that it's right and write a test that proves it's right; or we just accept that it's racy and document that.

I don't honestly know the Right Way here, but I suspect it is going to have to involve calling k_thread_abort() from an external manager thread or SVC handler.

rgundi commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	prompt "CMSIS RTOS v1 API"
+	default n
+	help
+	  This enables CMSIS RTOS v1 API support.
```
Sure. Will update.

rgundi commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	int
+	prompt "Maximum thread count in CMSIS RTOS application"
+	default 10
+	range 0 255
```
Well.. the limitation is because of the memory used to allocate the threads (using mempools/memslabs). I wanted to keep this uniform with POSIX.

rgundi commented on include/cmsis_rtos_v1/cmsis_os.h:
```diff
@@ -0,0 +1,703 @@
+/* ----------------------------------------------------------------------
+ * $Date:        5. February 2013
```
There's a newer set of APIs (CMSIS RTOS v2). However, Anas feels that's more for TEE and MPU scenarios and can be supported later once we have the support for v1.

rgundi commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -0,0 +1,40 @@
+	int
+	prompt "Maximum timer count in CMSIS application"
+	default 5
+	range 0 255
```
Wanted to keep this uniform with POSIX. The implementation is pretty similar.

rgundi commented:
@andyross : Thanks for the thorough review. I have replied for each of your comments and have made the changes ready. However, I need #8580 to get merged before i upload these changes (otherwise build will fail). Will keep you posted.

galak commented:
What's the general reason for adding this functionality?

nashif commented:
> What's the general reason for adding this functionality?

we have users moving from another RTOS that rely on this API.

rgundi commented:
@andyross  I have addressed your comments.
@andrewboie As discussed yesterday, have allocated k_thread statically thereby eliminating the usage of memslab or mempool (no free headache).

rgundi commented:
@nashif The shippable is failing due to 2 things here.
1) checkpatch - this fails for cmsis_os.h. We agreed the other day that this could be ignored and hence i am not worried about that.
2) Timeout - Some failure is happening in "build_ci" (see below). Not able to figure out what's happening here.
./scripts/sanitycheck ${SANITYCHECK_OPTIONS} --load-tests test_file.txt --subset ${MATRIX_BUILD}/${MATRIX_BUILDS} || ./scripts/sanitycheck ${SANITYCHECK_OPTIONS_RETRY} || ./scripts/sanitycheck ${SANITYCHECK_OPTIONS_RETRY} 

Apparently 401 out of 463 passed. It has probably failed in 402
401/463 native_posix samples/testing/integration/section.subsection PASSED

andyross commented on lib/cmsis_rtos_v1/cmsis_thread.c:
```diff
@@ -0,0 +1,177 @@
+{
+	void * (*fun_ptr)(void *) = arg3;
+
+	fun_ptr(arg1);
```
Why even bother with this wrapper function?  ARM arguments are passed in registers, so there's no reason not to just stuff the fun_ptr in arg3 into the k_thread_create() call and save a whole stack frame.

rgundi replied:
Well.. this is mainly because the 2 entrypoint function pointers are incompatible.  Zephyr is void (*)(void *, void *, void *) where as cmsis is void (* const)(const void *). I see it is mostly safe here but am just paranoid and would like to keep it the way it is.

andyross reviewed (approved):
I love the much better thread cleanup achieved by not having a struct needing cleanup at all! :)

Still think you're doing a bunch of senseless error checking trying to catch conditions that are not required to be runtime errors, but that's just code bloat in a wrapper library -- people who really cared wouldn't be using CMSIS at all.

galak commented:
> we have users moving from another RTOS that rely on this API.

Cool, would be neat if someone did something similar for FreeRTOS

rgundi committed:
lib/cmsis_rtos_v1: Implement support for mutex APIs

These APIs allow creating, acquiring, releasing and deleting
the mutexes.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of Mutex APIs

Some test programs to illustrate the usage of Mutex APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v1: Implement support for semaphore APIs

These APIs allow creating, acquiring, releasing and deleting
of semaphores.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of Semaphore APIs

Some test programs to illustrate the usage of Semaphore APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v1: Implement support for mempool APIs

These APIs allow creating, allocating and freeing
of mempools.

Note: "Mempool" in CMSIS actually means memslabs in Zephyr.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of mempool APIs

Some test programs to illustrate the usage of mempool APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@nashif Shippable is fine now except for the stylecheck and documentation. Both are because of cmsis_os.h and we decided the other day not to bother about it.
@andrewboie Please review the patch.

nashif commented:
add the header to EXCLUDE in the doxygen config. I do not think we should reference it this way, instead create documentation for the feature and reference the upstream documentation and API reference.

rgundi commented:
@nashif Addressed your suggestion. The shippable is failing only because of the checkpatch issue with cmsis_os.h. Everything else is clean now.

rgundi committed:
lib/cmsis_rtos_v1: Implement support for kernel APIs

These APIs allow for checking whether the kernel is initialized,
started or running.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
lib/cmsis_rtos_v1: Implement support for Mail Queue APIs

These APIs provide mail queue functionality like create, put and get.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of mailq APIs

Some test programs to illustrate the usage of mailq APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -0,0 +1,44 @@
+	  There's a limitation on the number of threads due to memory
+	  related constraints.
+
+config MAX_STACK_SIZE_THREAD
```
These are already under "if CMSIS_RTOS_V1_IPC". That is what you meant, right?

rgundi replied:
will change it to CMSIS_THREAD_MAX_STACK_SIZE.

rgundi commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -0,0 +1,44 @@
+	help
+	  Mention max stack size threads can be allocated in CMSIS RTOS application.
+
+config MAX_TIMER_COUNT
```
These are already under "if CMSIS_RTOS_V1_IPC". That is what you meant, right?

rgundi replied:
will change it to CMSIS_TIMER_MAX_COUNT.

dbkinder commented on doc/portability/portability.rst:
```diff
@@ -0,0 +1,12 @@
+.. _portability:
+
+Portability Guide
```
This title bothers me.  I think the document(s) in this section are about APIs from other RTOS systems that are provided by Zephyr to aid porting efforts from those RTOSes, right?  If so, a better title and description are warranted, maybe:

    .. _porting_to_zephyr:

    Porting Applications to Zephyr
    ##############################

    This section describes standard API interfaces provided by Zephyr
    specifically to ease porting applications originally written for other RTOS systems.

I'd also recommend not creating a new folder for these documents and just put them in the existing ``porting`` folder.

rgundi replied:
sure. i agree. i'll modify and resubmit.

nashif commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -0,0 +1,44 @@
+	help
+	  Mention maximum number of timers in CMSIS compliant application.
+
+config NUM_PREEMPT_PRIORITIES
```
ok, this is weird for an API to enforce number of priorities an RTOS should support. I see in v2 they have far more priorities than v1.

rgundi commented:
@nashif Addressed your comments.
@dbkinder Will address your comments tomorrow.

nniranjhana commented on tests/cmsis_rtos_v1/threads/thread_apis/src/main.c:
```diff
@@ -0,0 +1,136 @@
+	zassert_true(status == osOK, "Error setting thread priority\n");
+
+	/* Yield to thread of equal or higher priority */
+	status = osThreadYield();
```
AFAICS, there is no need to yield here, since just setting this thread's priority to low (as the `osThreadSetPriority` does just above) immediately gets thread2 (or the higher priority thread) to start executing on the next tick. Isn't it?

rgundi replied:
That's a good catch. Will modify the test. The previous osThreadSetPriority is not required.

nniranjhana commented on tests/cmsis_rtos_v1/threads/thread_apis/src/main.c:
```diff
@@ -0,0 +1,136 @@
+	 * its execution. It loops at the end and is terminated here.
+	 */
+	do {
+		status = osDelay(10);
```
`osDelay(100)` instead? Also, do we need to set status here? It is being overwritten by `osThreadTerminate` below anyway..

rgundi replied:
Will remove the status. will keep the delay to 10ms and modify the comment.

nniranjhana commented on tests/cmsis_rtos_v1/threads/thread_apis/src/main.c:
```diff
@@ -0,0 +1,136 @@
+	zassert_true(id2 != NULL, "Failed creating thread2\n");
+
+	do {
+		status = osDelay(100);
```
Did you mean to make use of status variable below? If not, can remove

rgundi replied:
yes. will remove.

nniranjhana commented on tests/cmsis_rtos_v1/threads/thread_apis/src/main.c:
```diff
@@ -0,0 +1,136 @@
+
+#define STACKSZ		512
+
+static int thread1_evidence;
```
`thread1_evidence` can be named better? The variable is tracking the flow between two threads, so something on that note? Also, a comment here for what the variables are for, would be helpful

rgundi replied:
renamed it to thread_yield_check.

nniranjhana commented on lib/cmsis_rtos_v1/cmsis_timer.c:
```diff
@@ -0,0 +1,137 @@
+
+static void zephyr_timer_wrapper(struct k_timer *timer);
+
+struct timer_obj {
```
Would the timer_obj struct be better off in a header file? I see that this is similar to what POSIX has; but since there is no `timer_obj` in CMSIS spec, is it okay to have these wrappers in user-exposed implementation files?

rgundi replied:
well.. i don't see a need to have a separate header file for just one structure. Also, i don't see any issue with wrappers being exposed in the implementation file. We are open source anyway :)

nniranjhana reviewed (commented):
Mostly just nitpicking in the added tests

rgundi commented:
@dbkinder Addressed your suggestions.

rgundi commented:
@nniranjhana Addressed your comments

rgundi committed:
lib/cmsis_rtos_v1: Implement support for Message Queue APIs

These APIs provide message queue functionality like create,
put and get.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of messageQ APIs

Some test programs to illustrate the usage of msgq APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@nashif The shippable is encountering a new failure (see below). This is apparently not due to the changes in this PR. Not quite sure what's happening here.

make[3]: Entering directory '/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/sanity-out/mimxrt1050_evk/tests/cmsis_rtos_v1/threads/thread_instance/portability.cmsis_rtos_v1'
[ 73%] Building C object zephyr/drivers/serial/CMakeFiles/drivers__serial.dir/uart_mcux_lpuart.c.obj
In file included from /home/buildslave/src/github.com/zephyrproject-rtos/zephyr/drivers/serial/uart_mcux_lpuart.c:11:0:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/ext/hal/nxp/mcux/drivers/imx/./fsl_lpuart.h: In function ‘LPUART_GetDataRegisterAddress’:
/home/buildslave/src/github.com/zephyrproject-rtos/zephyr/ext/hal/nxp/mcux/drivers/imx/./fsl_lpuart.h:500:30: error: ‘LPUART_Type {aka struct <anonymous>}’ has no member named ‘data’
     return (uint32_t) & (base->DATA);

rgundi commented:
@nashif As we discussed, I have added additional commits (from 66e7338 to 818b683) which address the API implementation and corresponding tests for mutex, semaphore, mempool, mailQ and messageQ.
@andrewboie @andyross @nniranjhana Please review only the new commits mentioned above.

dbkinder commented on doc/porting/cmsis_rtos_v1.rst:
```diff
@@ -0,0 +1,11 @@
+CMSIS RTOS v1
+##########################
+
+CMSIS expands into Cortex-M Software Interface Standard. CMSIS RTOS is a
```
A bit clearer would be to start this paragraph with:

    Cortex-M Software Interface Standard (CMSIS) RTOS is a
    vendor-independent ...

rgundi replied:
will make the change. Thanks David.

dbkinder reviewed (changes requested):
That works better, thanks.  One suggested change.

nniranjhana commented on lib/cmsis_rtos_v1/cmsis_semaphore.c:
```diff
@@ -0,0 +1,118 @@
+		return k_sem_count_get(semaphore);
+	}
+
+	return -1;
```
I'm not sure whether returning -1 here is right - could you check? In CMSIS doc, it says to return

> -1 in case of incorrect parameters

which is okay for the first two checks at the start; but in the case that the semaphore waiting period timed out, `k_sem_take` would return non-zero (EAGAIN) to `status`, indicating unavailability, and that should mean returning 0 rather -1 from `osSemaphoreWait`, since acc to doc again,

> The return value indicates the number of available tokens (the semaphore count value). If 0 is returned, then no semaphore was available.

rgundi replied:
That's correct. status will be 0 only if the semaphore was taken. So, in case of timeout, i should be returning 0 (indicating unavailability) instead of -1 (which is an error). Thanks for the catch.

nashif commented on lib/cmsis_rtos_v1/Kconfig:
```diff
@@ -0,0 +1,60 @@
+#
+# SPDX-License-Identifier: Apache-2.0
+#
+config CMSIS_RTOS_V1_IPC
```
What is the IPC for here? You probably copied that from the POSIX API which is also wrong. I would remove that and just call it CMSIS_RTOS_V1

rgundi replied:
Yes. Had copied from POSIX. Will remove.

nashif commented on lib/cmsis_rtos_v1/cmsis_mempool.c:
```diff
@@ -0,0 +1,73 @@
+	osPoolDef_t *osPool = (osPoolDef_t *)pool_id;
+	void *ptr;
+
+	if (k_mem_slab_alloc((struct k_mem_slab *)(osPool->pool), &ptr, 100)
```
Where is the 100 coming from? magic number?

rgundi replied:
i'll make it a macro. It's the timeout parameter and putting it in Kconfig will be an overkill.

rgundi commented:
@dbkinder @nashif @nniranjhana Addressed all your review comments.

dbkinder reviewed (approved):
+1 for docs, thanks.

spoorthik commented on lib/cmsis_rtos_v1/cmsis_mempool.c:
```diff
@@ -0,0 +1,75 @@
+		return NULL;
+	}
+
+	return ptr;
```
@rgundi : Do we need to return? I think we don't reach here, there is already return in both if and else case. Or did I miss understanding anything?

rgundi replied:
You are right. That return statement is not required. I'll correct that. Wonder why the compiler didn't catch this though. It should have said "unreachable statement" or something like that. @nashif Any idea why some warnings are apparently not enabled?

nashif replied:
we do not have such a warning, just remove it.

rgundi replied:
ok. will remove.

nashif commented:
Please group all tests into 1 or 2 tests in total, the tests you are adding per object are too small to justify a project on their own that would slow CI. This will generate ca. 900 builds when run with --all.

Basically what we need is one or two cmsis_rtos tests that verifies all features, this should be possible given the little code you have in some of those tests.

nniranjhana commented on tests/cmsis_rtos_v1/msgq/src/main.c:
```diff
@@ -0,0 +1,73 @@
+
+	/* Prepare and send 1st Message */
+	status = osMessagePut(message_id, data, osWaitForever);
+	zassert_true(status == osOK, "osMessagePut failure for Message1\n");
```
In all zasserts, you can remove the \n, it is not needed anymore as it is added by default

rgundi replied:
Yup. Thanks for pointing out.

rgundi committed:
lib/cmsis_rtos_v1: Implement support for Signal Event APIs

Signals are used to trigger execution states between threads.
These APIs provide functionalities like signal set, clear and
wait.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi committed:
tests/cmsis_rtos_v1: Introduce tests to make use of Signal APIs

Some test programs to illustrate the usage of Signal Event APIs.

Signed-off-by: Rajavardhan Gundi <rajavardhan.gundi@intel.com>

rgundi commented:
@nashif Will group all the tests under one test case (one file for each module instead of one folder which was the case earlier). Will shortly upload the changes.

rgundi commented:
@nashif @spoorthik @nniranjhana Addressed all your comments.

rgundi commented:
rebased and fixed an issue in Mailbox and MessageQs.
